### PR TITLE
DOSE-252 Kernel send all frees as part of end txg

### PR DESCRIPTION
**Pull request by grwilson**
This PR adds a linked list to maintain freed blocks so they can be processed at the end of the txg. When the list is processed, we make individual calls to the agent for each block. We may want to enhance this later by either supplying ranges of blocks to free or the entire list in one call. As a side-effect, this PR enables the pool checkpoint code path (I have not tested this) and may not be the ultimate implementation we would need for checking pointing object-based pools.

It is also worth noting that "deferred frees" feature (where we don't actually add the freed block to the allocatable range tree) is not fixed in the PR. So it's still possible that the agent may process the frees within the TXG_DEFER_SIZE window. This functionality would have to be moved to the agent in order for it to work the same way it does with block-based pools.

### CODE REVIEW

**Comment by grwilson** [comment]
> Is the point of this that if the agent crashes, the "free block" messages will be replayed?

Yes. If the agent crashes and we've not completed the end txg then we will replay the frees. I made the assumption that the agent is okay with receiving duplicate frees but maybe that is not true.

File: module/os/linux/zfs/vdev_object_store.c [hunk 1]
```diff
@@ -821,18 +845,28 @@ object_store_end_txg(vdev_t *vd, nvlist_t *config, uint64_t txg)
...
+		    nvbuf, nvlen);
+		fnvlist_pack_free(nvbuf, nvlen);
+
+		if (vos->vos_config != NULL)
+			fnvlist_free(vos->vos_config);
+		vos->vos_config = fnvlist_dup(config);
+	}
+
 	mutex_exit(&vos->vos_sock_lock);
 	agent_wait_serial(vos, VOS_SERIAL_END_TXG);
```
[the remark this answers is not in the document]
Comment (grwilson) [reply module/os/linux/zfs/vdev_object_store.c:863]: Yes, if we fail to send then the kernel should close connection causing the agent to fail when it tries to send and restart. The kernel will then wait to re-establish the connection and go through the normal resume logic. So the kernel's agent_resume logic will handle resending anything that was not acknowledged by the agent.

[the remark this answers is not in the document]
Comment (grwilson) [reply module/os/linux/zfs/vdev_object_store.c:863]: This is a code path where we could do something similar to the random failures in the agent. We could fail the send randomly as part of fault injection.

**Comment by ahrens** [comment]
> I made the assumption that the agent is okay with receiving duplicate frees but maybe that is not true.

As long as the duplicate frees come after a "open pool", it will work, because the agent will discard any of its own on-disk metadata (e.g. free block list) from the in-progress txg.  i.e. "free block A" ... (no "end txg done"; agent probably restarts) then "open pool" then "free block A".

**Review by don-brady — APPROVED** [review]

**Pull request merged**